### ISSUE TITLE
Can't create task that runs flutter pub run ...

### ISSUE THREAD

**Issue by JohnGalt1717** [Dart-Code/Dart-Code#3399]
**Describe the bug**
I'm trying to setup a task in VS Code that runs flutter pub run on a given directory (the workspace has multiple flutter projects and thus needs to run in a specific directory relative to the worksspace root folder.)

**To Reproduce**
```
  "tasks": [
    {
      "type": "flutter",
      "command": "flutter",
      "args": ["pub", "run", "localization_generator", "--output=./lib/l10n", "--input=./lib/l10n"],
      "label": "Regen Localization",

      "options": {
        "cwd": "${workspaceFolder}/libraries/localizations"
      },
      "problemMatcher": {
        "severity": "error",
        "pattern": {
          "regexp": "(?<=exit\\scode\\s)(\\d+)",
          "code": 1
        }
      }
    }
  ]
```

**Expected behavior**
This or some other variation should allow flutter pub run to execute on the given cwd instead of throwing an error that "run" doesn't exist.

**Versions (please complete the following information):**
 - VS Code version: 1.56.2
 - Dart extension version:  3.23.0
 - Dart/Flutter SDK version: 
 
Flutter 2.2.0 • channel beta • https://github.com/flutter/flutter.git
Framework • revision b22742018b (4 weeks ago) • 2021-05-14 19:12:57 -0700
Engine • revision a9d88a4d18
Tools • Dart 2.13.0

**Comment by DanTup** [comment]
This appears to be working as you'd like for me:

<img width="713" alt="Screenshot 2021-06-09 at 17 55 38" src="https://user-images.githubusercontent.com/1078012/121397106-f67c3f00-c94b-11eb-96f6-144cce0e7761.png">

Can you confirm how you're invoking the task? I gave it a unique `label` and then ran it from the `Run Task` command in the command palette and selected it from the list.

If that's not working, can you confirm what you see in the Terminal where it says "Executing task:"?

**Comment by JohnGalt1717** [comment]
It may be windows... the redacted part is just my initials without spaces or anything else. (4 characters)

**Comment by DanTup** [comment]
I've had no luck reproducing on Windows either, it seems to run fine (although it looks like Flutter or Pub are resetting the working directory once started which might be a problem for you, but I don't that's believe the cause of the error you're getting):

![Untitled](https://user-images.githubusercontent.com/1078012/121529825-af965400-c9f4-11eb-8eb7-adda1de638f0.png)

Does the built-in terminal work correctly for you? If you click the `+` on the Terminal tab, does it successfully create a terminal?

**Comment by DanTup** [comment]
Could you also try changing `"type": "flutter",` to `"type": "shell",` to see if that makes any difference?

**Comment by JohnGalt1717** [comment]
Yes, the terminal works fine and I can run the command if I go into the directory and just run it. I think the issue is that the command it appears to be executing doesn't have "flutter" on the front and is just "pub" with my example above.  as a result it can't find anything to run.

And yes the CWD is vital because the localization tool has to be in the right directory to be executed to generate the .dart file.

Changing it to shell and using flutter as the command, causes the OS to use the CWD + flutter to execute which then throws an error because that isn't where flutter.exe lives.

Basically to be able to automate the running of things with flutter with tasks it needs to:

1. cd {path}
2. run flutter using PATH.
3. Append all of the arguments to the flutter command.

Same would be true of dart.  I can't find any hack to make this work which blocks automating these processes that are standard (especially for localization but also for multiple build_runners that are watching on a mono-repro too as an example)

**Comment by DanTup** [comment]
>  I think the issue is that the command it appears to be executing doesn't have "flutter" on the front and is just "pub" with my example above. as a result it can't find anything to run.

Yeah, though I'm not sure what's really going on here. It seems to work correctly here and I can't really see any opportunity for it to not get `flutter` in the front.

Are you able to make a small repro project so I can test it with exactly the same setup? Be sure to commit everything (including the `.vscode` folder).

Otherwise, if I add some additional logging could you either run the extension from source, or I can make a build you can install that might help us track down what's going on?

**Comment by JohnGalt1717** [comment]
Litterally if you copy and paste the above code from the first post exactly into tasks.json and run it you get it.

If it worked properly you'd get a different error about localization_generator not being available.

**Comment by DanTup** [comment]
> Litterally if you copy and paste the above code from the first post exactly into tasks.json and run it you get it.

I did try that, along with several variations, but it executes fine for me. The working directory isn't as expected, but it's not failing to spawn the shell as it does in your message:

![image](https://user-images.githubusercontent.com/1078012/121543575-ef633880-ca00-11eb-911d-83c1b3c4885b.png)

I can look at why the `cwd` isn't set correctly, but that doesn't seem to be the same issue you're hitting, which is why I was hoping for either an exact project you can trigger this on (to remove the possibility that there's something else different between the projects we're trying with) or to add some additional logging.

**Comment by JohnGalt1717** [comment]
So I created a new flutter project, and put in the tasks exactly as above without alteration. And got the error.

I then though, wait, the cwd is wrong, I'll fix that, and removed the /localizations part. Worked fine then.  Then I moved the flutter project into a sub directory but kept the .vscode folder where it was and then updated the cwd in the tasks.json again, and volia, it works fine.

And explicitly in the terminal window the first time that it didn't work it didn't show flutter.bat on the command, and then after moving stuff around per above it did.

**Comment by DanTup** [comment]
@JohnGalt1717 I'm not sure I understand - your instructions sound the same as what I was doing (I didn't restart VS Code at all while trying to reprduce). Can you be more specific about what you think is causing it?

**Comment by JohnGalt1717** [comment]
There is something in vs code that if you create the flutter project in a sub directory in vs code while the root folder is open and then add the task.json with the cwd set that will cause the problem until you restart. As soon as you restart it works fine exactly as above.  And the error message in this scenario before restart is identical to what I posted.

If you create the flutter project in the root and don't set the cwd (which defaults to the workspace in that case) it's fine.
If you cdreate the flutter project in a sub folder without vs code open and then open vscode it will be fine too.

**Comment by DanTup** [comment]
@JohnGalt1717 can you give some detailed steps to repro? I tried this:

- Created an empty folder
- Opened the empty folder in VS Code
- In the built-in terminal, ran `flutter create foo` to create a project in a subfolder
- Copied your `task.json` above, but changed `cwd` to `${workspaceFolder}/foo/libraries/localizations` to reflect the nested project
- Created the `foo/libraries/localizations` folder
- Ran the task using the command palette

This all seemed to work as expected, although [Flutter seems to change the working directory](https://github.com/flutter/flutter/blob/c99ed373b3d254f9b10789ddf3771c3aa39e48ca/packages/flutter_tools/lib/src/runner/flutter_command.dart#L1398) ("Changing current working directory to: /Users/danny/Desktop/my_empty_folder/foo") when it runs - I don't think there's anything the extension here can do about that.

Is this different to what you're seeing?

**Comment by JohnGalt1717** [comment]
@DanTup 

1. Create a folder foo
2. cd into the foo
3. code .
3. from code terminal window flutter create bar
4. copy task.json from above into the .vscode folder in foo
5. You should now see .vscode and bar folders in vs code at the same level inside of foo which is open.
6. Now the above will occur until restart. (and it apparently periodically occurs on new startup)

**Comment by DanTup** [comment]
@JohnGalt1717 if I do that, I initially get an error like "no task provider for Flutter" which is because the extension isn't loaded (I think VS Code needs an additional activation event to fix that, I'll file an issue). If I open a .dart file to cause the extension to activate, then I see this:

<img width="1085" alt="Screenshot 2021-06-14 at 14 06 30" src="https://user-images.githubusercontent.com/1078012/121897069-e7b1d580-cd19-11eb-9926-d7393d0ce903.png">

This seems expected, as the path from the JSON above is not valid (not only does `libraries/localizations` not exist, but the project is also in a subfolder called `bar` that's not included in the `cwd`).

**Comment by JohnGalt1717** [comment]
@DanTup Sorry I used your directory names. Change "localization" to "bar" in the tasks.json and try again.

**Comment by DanTup** [comment]
@JohnGalt1717 do you mean change `cwd` to `"cwd": "${workspaceFolder}/bar"`? I did that but it seemed to run correctly (or at least, fail in the expected way - showing that it did run the command):

<img width="1193" alt="Screenshot 2021-06-14 at 14 17 11" src="https://user-images.githubusercontent.com/1078012/121898380-39a72b00-cd1b-11eb-85b4-2e71d73572bb.png">

I'll try on Windows again shortly in case it's different there.

**Comment by JohnGalt1717** [comment]
Yes.  And to be clear, when it works right the log shows flutter pub run and when it doesn't it only shows pub run with no "flutter" on the start of the command.

**Comment by DanTup** [comment]
Still no luck.

![Untitled](https://user-images.githubusercontent.com/1078012/121902615-63fae780-cd1f-11eb-84fd-17d2e1bd2775.png)

I still had to open a .dart file first to trigger activation of the extension (otherwise I get "no task provider registered for flutter") - I've filed https://github.com/microsoft/vscode/issues/126238 to try and improve that.

I presume you're on the latest versions of VS Code and the Dart extension? If I create a build that has additional logging, would you be able to try it out to capture some more info?

**Comment by JohnGalt1717** [comment]
Strange.

Yes, I'm using the latest (May 2021) version of vs code and the latest (according to the extensions tab) version of the Dart Extension.  I'm happy to use a build with additional logging, I'll just need instructions on how to install it.

**Comment by DanTup** [comment]
I've made a build that adds some logging. You can grab it from here:

https://github.com/Dart-Code/Dart-Code/releases/tag/v3.24.0-alpha.1

Installation instructions are on that page. Once installed (and Reloaded), get to the point where you think running the task will fail, and then:

- Run the **Dart: Capture Extension Logs** command from VS Code
- Run the task
- Click **Cancel** on the logging notification

It should include some logging like this:

```
[11:08:56 AM] [General] [Info] Resolving task...
[11:08:56 AM] [General] [Info] URI from scope is /Users/danny/Desktop/my_empty_folder
[11:08:56 AM] [General] [Info] Creating TaskExecution for flutter...
[11:08:56 AM] [General] [Info] 
			Returning a vs.ProcessExecution:
			  program: /Users/danny/Dev/Google/flutter/bin/flutter
			  args: pub,run,localization_generator,--output=./lib/l10n,--input=./lib/l10n
			  options: {"cwd":"/Users/danny/Desktop/my_empty_folder","env":{"FLUTTER_HOST":"VSCode","PUB_ENVIRONMENT":"vscode.dart-code"}}
```
The last line is showing exactly what we're returning to VS Code for it to run. Note: the cwd may be incorrect, but VS Code doesn't seem to give us the cwd you've defined in the task, it appears to overwrite what we return here with it though.

**Comment by JohnGalt1717** [comment]
Sorry, I'll be looking at this on Monday. Was away on vacation.

**Comment by Panama202** [comment]
Hola

**Comment by JohnGalt1717** [comment]
Sorry for the delay.  This now works reliably and every time with the latest build:

    {
      "type": "flutter",
      "command": "flutter",
      "args": ["pub", "run", "localization_generator", "--output=./lib/l10n", "--input=./l10n"],
      "label": "Regen Localization",
      "detail": "Builds the localization Dart files from json",
      "options": {
        "cwd": "${workspaceFolder}/libraries/localizations"
      },
      "presentation": {
        "echo": true,
        "reveal": "always",
        "focus": true,
        "panel": "new",
        "showReuseMessage": false,
        "clear": true
      }
    },
